addons/netfox/rollback/rollback-synchronizer.gd

```diff
@@ -192,105 +231,168 @@ func _after_tick(_delta, _tick):
-	var sender = multiplayer.get_remote_sender_id()
-	var sanitized = {}
+func _submit_inputs(input: Dictionary, tick: int):
+	var sender_id: int = multiplayer.get_remote_sender_id()
+	var sanitized: Dictionary = {}
+
 	for property in input:
 		var pe = _property_cache.get_entry(property)
 		var value = input[property]
-		var input_owner = pe.node.get_multiplayer_authority()
-		
-		if input_owner != sender:
+		var input_owner_id = pe.node.get_multiplayer_authority()
+
+		if input_owner_id != sender_id:
 			_logger.warning("Received input for node owned by %s from %s, sender has no authority!" \
-				% [input_owner, sender])
+				% [input_owner_id, sender_id])
 			continue
-		
+
 		sanitized[property] = value
-	
+
 	if sanitized.size() > 0:
 		for property in sanitized:
 			for i in range(0, sanitized[property].size()):
 				var t = tick - i
 				var old_input = _inputs.get(t, {}).get(property)
 				var new_input = sanitized[property][i]
-				
+
 				if old_input == null:
 					# We received an array of current and previous inputs, merge them into our history.
 					_inputs[t] = _inputs.get(t, {})
 					_inputs[t][property] = new_input
 					_earliest_input = min(_earliest_input, t)
 	else:
-		_logger.warning("Received invalid input from %s for tick %s for %s" % [sender, tick, root.name])
+		_logger.warning("Received invalid input from %s for tick %s for %s" % [sender_id, tick, root.name])
 
 @rpc("any_peer", "unreliable_ordered", "call_remote")
-func _submit_state(state: Dictionary, tick: int):
-	if tick > NetworkTime.tick:
-		# This used to be weird, but is now expected due to estimating remote time
-		# push_warning("Received state from the future %s / %s - adding nonetheless" % [tick, NetworkTime.tick])
-		pass
-	
-	if tick < NetworkTime.tick - NetworkRollback.history_limit and _latest_state >= 0:
+func _submit_state(received_state: Dictionary, tick: int):
+	if tick < NetworkTime.tick - NetworkRollback.history_limit and _latest_state_tick >= 0:
 		# State too old!
 		_logger.error("Received state for %s, rejecting because older than %s frames" % [tick, NetworkRollback.history_limit])
 		return
-
-	var sender = multiplayer.get_remote_sender_id()
-	var sanitized = {}
-	for property in state:
-		var pe = _property_cache.get_entry(property)
-		var value = state[property]
-		var state_owner = pe.node.get_multiplayer_authority()
 		
-		if state_owner != sender:
+	#if (_states.has(tick)):
+		#_logger.warning("Received state %s AGAIN" % tick)
+		#return
+
+	# It is guaranteed that the first states are full, until the client sends acks to the server
+	# which enables the server to start sending diffs
+	if (not _has_received_full_state):
+		_has_received_full_state = true
+		_receive_full_state_ack.rpc_id(1, tick)
+
+	if (received_state.size() > 0):
+		set_received_state(received_state, tick, multiplayer.get_remote_sender_id())
+	elif (NetworkRollback.enable_state_diffs):
+		if (_states.has(tick - 1)):
+			_states[tick] = _states[tick - 1]
```

Comment:
Why do this? At this point, `_states[tick]` should already be set by `set_received_state`, which this then overrides fully.

Reply:
Because if a diff contains 0 properties (that means there was no change in the properties from previous tick), the new state must simply be a duplicate of the previous one.